shared/src/main/scala/async/retry.scala

```diff
@@ -0,0 +1,163 @@
+package gears.async
+
+import gears.async.Async
+import scala.concurrent.duration._
+import scala.util.Try
+import scala.util.control.NonFatal
+import scala.util.Failure
+import scala.util.Success
+import gears.async.AsyncOperations.sleep
+import scala.util.Random
+import java.util.concurrent.TimeoutException
+import gears.async.Retry.Delay
+
+/** Utility class to perform asynchronous actions with retrying policies on exceptions.
+  *
+  * See [[Retry]] companion object for common policies as a starting point.
+  */
+case class Retry(
+    retryOnSuccess: Boolean = false,
+    maximumFailures: Option[Int] = None,
+    delay: Delay = Delay.none
+):
+  /** Runs [[body]] with the current policy in its own scope, returning the result or the last failure as an exception.
+    */
+  def apply[T](op: => T)(using Async, AsyncOperations): T =
+    @scala.annotation.tailrec
+    def loop(failures: Int, lastDelay: Duration): T =
+      Try(op) match
```

Comment:
It would be nice, but I do want to maintain compatibility with which exceptions to be wrapped and caught by `Try`. We could probably move it to catching `NonFatal` though.